models/stock_picking.py

```diff
@@ -322,14 +324,100 @@ def update_picking(
             # mark_as_done the stock.move.lines
             mls_done = move_lines.mark_as_done(**values)
 
+            if real_time_update:
+                picking = self._real_time_update(mls_done)
+                validate = True
+
         if force_validate or validate:
-            if self.move_line_ids.get_lines_todo() and not create_backorder:
+            if picking.move_line_ids.get_lines_todo() and not create_backorder:
                 raise ValidationError(
                         _('Cannot validate transfer because there'
                           ' are move lines todo'))
             # by default action_done will backorder the stock.move.lines todo
             # validate stock.picking
-            self.action_done() # old do_transfer
+            picking.action_done() # old do_transfer
+
+    def _requires_backorder(self, mls):
+        # if we
+        #   have all the moves
+        #   have all the move_lines within them
+        #   are not waiting for any previous moves to complete
+        # then we dont need to backorder
+        mls_moves = mls.mapped('move_id')
+        for move in self.move_lines:
+            if move not in mls_moves:
+                break
+            if not move.move_line_ids == mls.filtered(lambda x: x.move_id == move):
+                break
+            if move.move_orig_ids.filtered(lambda x: x.state not in ('done', 'cancel')):
+                break
+        else:
+            return False
+        return True
+
+    def _create_backorder(self, mls=None):
+        """ Creates a backorder pick from self and a subset of
+            stock.move.lines are then moved into it.
+            the move from which the move lines have been transfers
+            has the ordered_qty decrimented by the amount in the
+            transftered lines.
+        """
+        Move = self.env['stock.move']
+        # Based on back order creation in sock_move._action_done
+        self.ensure_one()
+
+        if mls is None:
+            # or done & canceled?
+            mls = self.move_lines.filtered(lambda x: x.state == 'done')
+        # make empty set
+        new_moves = Move.browse()
+        for current_move in mls.mapped('move_id'):
+            current_mls = mls.filtered(lambda x: x.move_id == current_move)
+
+            if current_mls == current_move.move_line_ids and \
+               not current_move.move_orig_ids.filtered(
+                            lambda x: x.state not in ('done', 'cancel')):
+                bk_move = current_move
```

Comment:
You should define `bk_move` to an empty / null value out of the `if`, specially because you're in a loop; this is very dangerous.